src/BaseMailer.php

```diff
@@ -0,0 +1,284 @@
+     *   In this case, the text body will be generated by applying `strip_tags()` to the HTML body.
+     * - an array with 'html' and/or 'text' elements. The 'html' element refers to the view name
+     *   for rendering the HTML body, while 'text' element is for rendering the text body. For example,
+     *   `['html' => 'contact-html', 'text' => 'contact-text']`.
+     * - null, meaning the message instance will be returned without body content.
+     *
+     * The view to be rendered can be specified in one of the following formats:
+     *
+     * - a relative view name (e.g. "contact") located under [[viewPath]].
+     *
+     * @param array $params the parameters (name-value pairs) that will be extracted and made available in the view file.
+     * @return MessageInterface message instance.
+     */
+    public function compose($view = null, array $params = []): MessageInterface
+    {
+        $message = $this->createMessage();
+        if ($view === null) {
+            return $message;
+        }
+
+        $this->composer->compose($message, $view, $params);
+
+        return $message;
+    }
+
+    /**
+     * Creates a new message instance.
+     * @return MessageInterface message instance.
+     */
+    protected function createMessage(): MessageInterface
+    {
+        return $this->messageFactory->create($this);
+    }
+
+    /**
+     * Sends the given email message.
+     * This method will log a message about the email being sent.
+     * If [[useFileTransport]] is true, it will save the email as a file under [[fileTransportPath]].
+     * Otherwise, it will call [[sendMessage()]] to send the email to its recipient(s).
+     * Child classes should implement [[sendMessage()]] with the actual email sending logic.
+     * @param MessageInterface $message email message instance to be sent
+     * @return bool whether the message has been sent successfully
+     */
+    public function send(MessageInterface $message): bool
+    {
+        if (!$this->beforeSend($message)) {
+            return false;
+        }
+
+        $address = $message->getTo();
+        if (is_array($address)) {
+            $address = implode(', ', array_keys($address));
+        }
+        $this->logger->info('Sending email "' . $message->getSubject() . '" to "' . $address . '"');
+
+        if ($this->useFileTransport) {
+            $isSuccessful = $this->saveMessage($message);
+        } else {
+            $isSuccessful = $this->sendMessage($message);
+        }
+        $this->afterSend($message, $isSuccessful);
+
+        return $isSuccessful;
+    }
+
+    /**
+     * Sends multiple messages at once.
+     *
+     * The default implementation simply calls [[send()]] multiple times.
+     * Child classes may override this method to implement more efficient way of
+     * sending multiple messages.
+     *
+     * @param array $messages list of email messages, which should be sent.
+     * @return int number of messages that are successfully sent.
```

Comment:
The value returned isn't useful to me as a developer. I'd prefer it to be array of exceptions per message or something so I can retry failed ones or provide a report on what was failed.